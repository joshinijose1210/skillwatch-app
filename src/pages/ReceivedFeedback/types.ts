export interface ReceivedFeedbackProps {
    action?: string;
}

export type ReceivedFeedbackTypes = { [key: string]: any }[];
