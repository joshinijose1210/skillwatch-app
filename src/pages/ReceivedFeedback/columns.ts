import { DateFormatter, FeedbackTagFormatter, FullNameReceivedFeedbackFormatter, ViewFeedbackButton } from '@components';
import { TableColumnConfig } from '@medly-components/core';

export const ReceivedFeedbackCols: TableColumnConfig[] = [
    { title: 'Date', field: 'date', sortable: true, component: DateFormatter, fraction: 0.4 },
    { title: 'Feedback From', field: 'empFirstName', component: FullNameReceivedFeedbackFormatter, fraction: 0.5 },
    { title: 'Feedback Type', field: 'feedbackType', fraction: 0.5, component: FeedbackTagFormatter },
    { title: 'Action', field: 'action', component: ViewFeedbackButton, fraction: 0.4 }
];
