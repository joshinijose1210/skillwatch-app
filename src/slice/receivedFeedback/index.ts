import { receivedFeedbackSlice } from './receivedFeedback';
export const { updateReceivedFeedbackFilter, resetReceivedFeedbackFilter } = receivedFeedbackSlice.actions;
export default receivedFeedbackSlice.reducer;
export * from './types';
