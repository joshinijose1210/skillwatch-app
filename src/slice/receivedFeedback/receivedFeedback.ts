import { createSlice } from '@reduxjs/toolkit';
import { ReceivedFeedbackState } from './types';

export const initialState: ReceivedFeedbackState = {
    activeFeedbackType: -1,
    activeFromEmployees: []
};

export const receivedFeedbackSlice = createSlice({
    name: 'receivedFeedback',
    initialState,
    reducers: {
        updateReceivedFeedbackFilter(state, { payload }) {
            return {
                ...state,
                ...payload
            };
        },
        resetReceivedFeedbackFilter() {
            return initialState;
        }
    }
});
