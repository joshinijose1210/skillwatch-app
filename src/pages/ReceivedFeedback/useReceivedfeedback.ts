import { routeConstants } from '@constants';
import { useAppSelector } from '@slice';
import { updateReceivedFeedbackFilter } from '@slice/receivedFeedback';
import { useGetEmployeeListQuery, useGetReceivedFeedbacksQuery, useGetTagsListQuery } from '@slice/services';
import { useCallback, useState } from 'react';
import { useDispatch } from 'react-redux';
import { useNavigate, useParams } from 'react-router-dom';

export const useReceivedFeedback = () => {
    const dispatch = useDispatch();
    const navigate = useNavigate();
    const { ActivePage } = useParams();
    const [page, setPage] = useState<number>(1);
    const { reviewCycleList } = useAppSelector(state => state.reviewCycleFilter);
    const userDetails = useAppSelector(state => state.user);
    const receivedFeedbackState = useAppSelector(state => state.receivedFeedback);
    const selectedTag = receivedFeedbackState.activeFeedbackType;
    const selectedEmployees = receivedFeedbackState.activeFromEmployees;
    const [sortOrder, setSortOrder] = useState('desc');

    const handleOrderChange = (order: string) => {
        setSortOrder(order);
        navigate(`${routeConstants.receivedFeedback}/${1}`);
    };
    const handlePageChange = (page: number) => {
        setPage(page);
        navigate(`${routeConstants.receivedFeedback}/${page}`);
    };
    const {
        data,
        isFetching: isLoading,
        isError
    } = useGetReceivedFeedbacksQuery(
        {
            path: 'received/',
            params: [
                { name: 'page', value: selectedEmployees.length == 0 ? ActivePage ?? page : 1 },
                { name: 'limit', value: 10 },
                { name: 'organisationId', value: userDetails.organisationId },
                { name: 'feedbackFromId', value: selectedEmployees.join(',') || -99 },
                { name: 'feedbackToId', value: userDetails.id },
                { name: 'sortBy', value: sortOrder === 'desc' ? 'dateDesc' : 'dateAsc' },
                { name: 'feedbackTypeId', value: selectedTag || -99 },
                { name: 'reviewCycleId', value: reviewCycleList }
            ]
        },
        { refetchOnMountOrArgChange: true }
    );

    const { data: tags, isSuccess: isTagsListReceived } = useGetTagsListQuery({});

    const handleDropdownChange = useCallback(
        (value: number) => {
            dispatch(
                updateReceivedFeedbackFilter({
                    activeFeedbackType: value
                })
            );
            navigate(`${routeConstants.receivedFeedback}/${1}`);
        },
        [navigate]
    );

    const handleSelectEmployee = (value: any) => {
        dispatch(
            updateReceivedFeedbackFilter({
                activeFromEmployees: value
            })
        );
        navigate(`${routeConstants.receivedFeedback}/${1}`);
    };

    const { data: employees, isSuccess: isEmployeeListReceived } = useGetEmployeeListQuery(
        {
            path: '',
            params: [
                { name: 'teamId', value: -99 },
                { name: 'designationId', value: -99 },
                { name: 'organisationId', value: userDetails.organisationId },
                { name: 'roleId', value: -99 },
                { name: 'searchText', value: '' },
                { name: 'departmentId', value: -99 }
            ]
        },
        { refetchOnMountOrArgChange: true }
    );

    const employeeList = useCallback(() => {
        let filteredList = [];
        if (isEmployeeListReceived) {
            filteredList = employees?.totalEmployees ? employees?.employees.filter((employee: any) => employee.status === true) : [];
        }
        let list = filteredList
            ? filteredList.map((employee: any) => {
                  return {
                      value: employee.id,
                      label: `${employee.firstName} ${employee.lastName} (${employee.employeeId})`
                  };
              })
            : [];
        list = list.filter((employee: any) => employee.value !== userDetails.id);

        list.sort((a: any, b: any) => a.label.localeCompare(b.label));
        return list;
    }, [employees, isEmployeeListReceived, userDetails.id]);

    const tagsList = isTagsListReceived
        ? [
              { value: -99, label: 'All' },
              ...tags?.map((tag: any) => {
                  return { value: tag.feedbackTypeId, label: tag.feedbackType };
              })
          ]
        : [];

    return {
        isLoading,
        employeeList,
        handleSelectEmployee,
        selectedEmployees,
        tagsList,
        handleDropdownChange,
        selectedTag,
        isError,
        data,
        ActivePage,
        handlePageChange,
        handleOrderChange
    };
};
