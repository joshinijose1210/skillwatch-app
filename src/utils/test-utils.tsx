import { ToastContainer } from '@medly-components/core';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import { store as rtkStore, setupStore } from '@slice';
import { RenderOptions, RenderResult, render } from '@testing-library/react';
import { defaultTheme } from '@theme';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import reduxMockStore from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';
import { userMockState } from '../mocks/userMockState';
import { initialState } from '../slice';

interface CustomOptions extends RenderOptions {
    initialState?: any;
    initialEntries?: any;
}

interface WithThemeProviderProps {
    children?: React.ReactNode;
}

const mockStore = reduxMockStore(),
    store = mockStore({
        ...initialState,
        user: { ...initialState.user, groups: ['admin'] }
    });

const WithThemeProvider: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <ThemeProvider theme={defaultTheme}>
        <>
            <ToastContainer position="top-end" />
            {children}
        </>
    </ThemeProvider>
);

const WithStore: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <WithThemeProvider>{children}</WithThemeProvider>
    </Provider>
);

const WithRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <MemoryRouter>
        <WithThemeProvider>{children}</WithThemeProvider>
    </MemoryRouter>
);

const WithStoreAndRouter: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
    <Provider store={store}>
        <MemoryRouter>
            <WithThemeProvider>{children}</WithThemeProvider>
        </MemoryRouter>
    </Provider>
);

export const RTKRender = (ui: React.ReactElement, { initialState = userMockState, ...renderOptions }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);

    const RtkWrapper: React.FunctionComponent<WithThemeProviderProps> = ({ children }) => (
        <Provider store={store}>
            <MemoryRouter initialEntries={renderOptions.initialEntries}>
                <WithThemeProvider>{children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
    return { rtkStore, ...render(ui, { wrapper: RtkWrapper, ...renderOptions }) };
};
const customRender = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithThemeProvider, ...options });

export const renderWithStore = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStore, ...options });

export const renderWithRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithRouter, ...options });

export const renderWithStoreAndRouter = (ui: React.ReactElement, options?: RenderOptions): RenderResult =>
    render(ui, { wrapper: WithStoreAndRouter, ...options });

export const wrapper = (props: { children: React.ReactNode }, { initialState = userMockState }: CustomOptions = {}) => {
    const store = setupStore(initialState);
    setupListeners(store.dispatch);
    return (
        <Provider store={store}>
            <MemoryRouter>
                <WithThemeProvider>{props.children}</WithThemeProvider>
            </MemoryRouter>
        </Provider>
    );
};

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render, store };
