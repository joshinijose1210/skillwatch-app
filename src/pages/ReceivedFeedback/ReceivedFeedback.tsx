import { StyledMultiSelect, StyledSingleSelect } from '@common';
import { PageContent } from '@components';
import { CustomTable } from '@components/reusableComponents/CustomTable/CustomTable';
import ListHeader from '@components/reusableComponents/ListHeader';
import { StyledWrapper } from '@pages/EmployeeManagement/EmployeeManagement.styled';
import { FC } from 'react';
import { ReceivedFeedbackCols as columns } from './columns';
import { ReceivedFeedbackProps } from './types';
import { useReceivedFeedback } from './useReceivedfeedback';
import TooltipDropdown from '@components/reusableComponents/TooltipDropdown';
import { OptionsType } from '@components/reusableComponents/TooltipDropdown/types';

export const ReceivedFeedback: FC<ReceivedFeedbackProps> = () => {
    const {
        data,
        isLoading,
        employeeList,
        handleSelectEmployee,
        selectedEmployees,
        tagsList,
        handleDropdownChange,
        selectedTag,
        isError,
        ActivePage,
        handlePageChange,
        handleOrderChange
    } = useReceivedFeedback();

    return (
        <PageContent>
            <ListHeader title="Received Feedback" />
            <StyledWrapper>
                <StyledMultiSelect
                    options={employeeList()}
                    variant="outlined"
                    placeholder="Select Employee Name"
                    label="Feedback From"
                    data-testid="feedbackFromDropDown"
                    onChange={value => handleSelectEmployee(value)}
                    values={selectedEmployees}
                    size="M"
                    minWidth="25rem"
                />
                <TooltipDropdown
                    dataIds={['feedback from-input', 'feedback from-count-chip']}
                    values={
                        selectedEmployees?.length && employeeList().length
                            ? employeeList().filter(
                                  (item: OptionsType) =>
                                      typeof item.value === 'number' && (selectedEmployees as number[]).includes(item.value)
                              )
                            : []
                    }
                />
                <StyledSingleSelect
                    options={tagsList}
                    label="Feedback Type"
                    data-testid="feedbackTypeDropDown"
                    placeholder="Select Feedback Type"
                    variant="outlined"
                    size="M"
                    onChange={value => handleDropdownChange(value)}
                    value={selectedTag}
                    minWidth="25rem"
                />
                <TooltipDropdown
                    dataIds={['feedback type-input']}
                    values={selectedTag && tagsList.length ? tagsList.filter((item: OptionsType) => selectedTag === item.value) : []}
                />
            </StyledWrapper>
            <CustomTable
                columns={columns}
                count={data?.totalFeedbacks || 0}
                data={isError ? [] : data?.feedbacks?.map((feedback: any) => ({ ...feedback, feedbackCategory: 'received' })) || []}
                isLoading={isLoading}
                activePage={parseInt(ActivePage || '1')}
                handlePageChange={handlePageChange}
                setSortOrder={handleOrderChange}
            />
        </PageContent>
    );
};
