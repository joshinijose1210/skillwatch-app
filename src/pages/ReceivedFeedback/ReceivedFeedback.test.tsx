import { RTKRender, cleanup, fireEvent, screen, waitFor } from '@utils/test-utils';
import { ReceivedFeedback } from './ReceivedFeedback';
import MockAdapter from 'axios-mock-adapter';
import { apiInstance } from '@utils';
import { receivedFeedbackMockData } from '@mocks/receivedFeedback';

const mock = new MockAdapter(apiInstance, { onNoMatch: 'throwException' });
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
    useLocation: jest.fn()
}));

beforeAll(() => {
    const { url, data } = receivedFeedbackMockData.received;
    mock.onGet(url).replyOnce(200, data);
    const { url: employeeListUrl, data: EmployeeList } = receivedFeedbackMockData.employees;
    mock.onGet(employeeListUrl).replyOnce(200, EmployeeList);
    const { url: feedbackTypeUrl, data: feedbackTypeList } = receivedFeedbackMockData.feedbackType;
    mock.onGet(feedbackTypeUrl).replyOnce(200, feedbackTypeList);
});

beforeEach(() => {
    cleanup();
});

afterEach(mock.reset);

describe('Received feedback component', () => {
    it('should take snapshot', () => {
        const { container } = RTKRender(<ReceivedFeedback />);
        expect(container).toMatchSnapshot();
    });

    it('should render propely', async () => {
        const { url, data } = receivedFeedbackMockData.received;
        mock.onGet(url).replyOnce(200, data);
        const { url: feedbackTypeUrl, data: feedbackTypeList } = receivedFeedbackMockData.feedbackType;
        mock.onGet(feedbackTypeUrl).replyOnce(200, feedbackTypeList);
        RTKRender(<ReceivedFeedback />);
        await waitFor(() => screen.findAllByText('View'), { timeout: 3000 });
        expect(screen.getByText('Received Feedback')).toBeInTheDocument();
        expect(screen.getByText('A D (SR0067)')).toBeInTheDocument();
    });

    it('should show no result when api fails', async () => {
        RTKRender(<ReceivedFeedback />);
        await screen.findByText('No result');
    });

    it('should navigate when click on view', async () => {
        const { url, data } = receivedFeedbackMockData.received;
        mock.onGet(url).replyOnce(200, data);
        const { url: feedbackTypeUrl, data: feedbackTypeList } = receivedFeedbackMockData.feedbackType;
        mock.onGet(feedbackTypeUrl).replyOnce(200, feedbackTypeList);
        RTKRender(<ReceivedFeedback />);
        await waitFor(() => screen.findAllByText('View'), { timeout: 3000 });
        fireEvent.click(screen.getAllByText('View')[0]);
    });

    it('should filter list when filter is selected', async () => {
        const { url, url2, url3, data, filterData } = receivedFeedbackMockData.received;
        mock.onGet(url).replyOnce(200, data).onGet(url2).replyOnce(200, filterData).onGet(url3).replyOnce(200, {});
        const { url: employeeListUrl, data: EmployeeList } = receivedFeedbackMockData.employees;
        mock.onGet(employeeListUrl).replyOnce(200, EmployeeList).onGet(employeeListUrl).reply(200, EmployeeList);
        const { url: feedbackTypeUrl, data: feedbackTypeList } = receivedFeedbackMockData.feedbackType;
        mock.onGet(feedbackTypeUrl).replyOnce(200, feedbackTypeList);
        RTKRender(<ReceivedFeedback />);
        await waitFor(() => screen.findAllByText('View'), { timeout: 3000 });
        fireEvent.change(screen.getByPlaceholderText('Select Employee Name'), { target: { value: 'M P (SR0023)' } });
        fireEvent.click(screen.getByText('M P (SR0023)'));
        await waitFor(() => screen.findAllByText('View'), { timeout: 3000 });
        expect(screen.getByText('B E (SR0023)')).toBeInTheDocument();
        fireEvent.click(screen.getByTestId('feedbackTypeDropDown'));
        await screen.findByText('Improvement');
        fireEvent.click(screen.getByText('Improvement'));
        await screen.findByText('No result');
    });

    it('should pagination work properly', async () => {
        const { url, pageUrl, data } = receivedFeedbackMockData.received;
        mock.onGet(url).replyOnce(200, data).onGet(pageUrl).replyOnce(200, {});
        RTKRender(<ReceivedFeedback />);
        await waitFor(() => screen.findAllByText('View'), { timeout: 3000 });
        fireEvent.click(screen.getByText(2));
        await screen.findByText('No result');
    });

    it('should sorting work propely', async () => {
        const { url, sortUrl, data } = receivedFeedbackMockData.received;
        mock.onGet(url).replyOnce(200, data).onGet(sortUrl).replyOnce(200, data);
        RTKRender(<ReceivedFeedback />);
        await waitFor(() => screen.findAllByText('View'), { timeout: 3000 });
        const spanElement = screen.getByText('Date');
        fireEvent.click(spanElement);
        await waitFor(() => screen.findAllByText('View'), { timeout: 3000 });
    });
});
