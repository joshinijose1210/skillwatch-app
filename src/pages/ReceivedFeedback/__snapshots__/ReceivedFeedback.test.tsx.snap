// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Received feedback component should take snapshot 1`] = `
<div>
  <main
    class="sc-1ehavxz-0 dDyhBJ"
  >
    <div
      class="sc-h70dbj-0 fQmYvl row expanded space-between align-center marginBottom-2"
    >
      <div
        class="sc-w8e7z6-2"
      >
        <h3
          class="sc-1ntgl3l-0 kcLBOi sc-w8e7z6-0 ibnmt"
          textalign="initial"
          textcolor="inherit"
          textvariant="h3"
          textweight="Medium"
        >
          Received Feedback
        </h3>
      </div>
    </div>
    <div
      class="sc-1h7claq-0 jdWERm"
    >
      <div
        class="sc-11tbpdo-0 fpmwxR sc-tkiih6-14 gSveWN"
        data-testid="feedbackFromDropDown"
        id="feedbackfrom-wrapper"
        minwidth="25rem"
        variant="outlined"
      >
        <div
          class="sc-17d4s39-0 UfceN"
          id="feedbackfrom-input-wrapper"
          minwidth="25rem"
        >
          <div
            class="sc-19px6zp-0 BrSGZ"
            filled="[object Object]"
            fusion="[object Object]"
            minrows="1"
            outlined="[object Object]"
            textvariant="[object Object]"
            variant="outlined"
          >
            <div
              class="sc-lhr232-0 btNOKw"
              variant="outlined"
            >
              <input
                aria-describedby="feedbackfrom-helper-text"
                autocomplete="off"
                class="sc-hb99wz-0 jaEbhp"
                errortext=""
                id="feedbackfrom-input"
                inputsize="M"
                placeholder="Select Employee Name"
                type="text"
                value=""
                variant="outlined"
              />
              <label
                class="sc-1du51u0-0 eMZrST"
                for="feedbackfrom-input"
                inputwidth="0"
                variant="outlined"
              >
                Feedback From
              </label>
            </div>
            <div
              class="sc-4vbg19-0 hFRmRP"
              id="feedbackfrom-count"
            >
              <svg
                class="sc-1pf3x8a-0 klDMsE"
                fill="none"
                height="1em"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12.792 14.97a1 1 0 0 1-1.584 0l-3.165-4.11a1 1 0 0 1 .793-1.61h6.328a1 1 0 0 1 .793 1.61l-3.165 4.11z"
                  fill="#000"
                />
              </svg>
            </div>
          </div>
          
        </div>
      </div>
      
      <div
        class="sc-1hcsztp-0 fKSUGy sc-tkiih6-13 cecHDR"
        id="feedback type-wrapper"
        minwidth="25rem"
        variant="outlined"
      >
        <div
          class="sc-17d4s39-0 UfceN"
          id="feedback type-input-wrapper"
          minwidth="25rem"
        >
          <div
            class="sc-19px6zp-0 BrSGZ"
            filled="[object Object]"
            fusion="[object Object]"
            minrows="1"
            outlined="[object Object]"
            textvariant="[object Object]"
            variant="outlined"
          >
            <div
              class="sc-lhr232-0 btNOKw"
              variant="outlined"
            >
              <input
                aria-describedby="feedback type-helper-text"
                autocomplete="off"
                class="sc-hb99wz-0 jaEbhp"
                data-testid="feedbackTypeDropDown"
                data-tooltip-id="feedback type-input"
                errortext=""
                id="feedback type-input"
                inputmode="none"
                inputsize="M"
                placeholder="Select Feedback Type"
                type="text"
                value=""
                variant="outlined"
              />
              <label
                class="sc-1du51u0-0 eMZrST"
                for="feedback type-input"
                inputwidth="0"
                variant="outlined"
              >
                Feedback Type
              </label>
            </div>
            <svg
              class="sc-1pf3x8a-0 klDMsE sc-dynlvq-0 bdmviR"
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12.792 14.97a1 1 0 0 1-1.584 0l-3.165-4.11a1 1 0 0 1 .793-1.61h6.328a1 1 0 0 1 .793 1.61l-3.165 4.11z"
                fill="#000"
              />
            </svg>
          </div>
          
        </div>
      </div>
      
    </div>
    <div
      class="sc-8tbhpn-3"
    >
      <div
        class="sc-1msmxxr-0 iWEzjW"
      />
      <table
        actions=""
        class="sc-1msmxxr-1 jwFsPF sc-8tbhpn-2 kCDmJI"
        columns="[object Object],[object Object],[object Object],[object Object]"
        defaultactivepage="1"
        defaultsortorder="desc"
        itemsperpage="10"
        rowclickdisablekey=""
        tabindex="-1"
        totalitems="0"
      >
        <thead
          class="sc-uxzsxp-0 haNltq"
        >
          <tr
            class="sc-1jr91bg-0 hJzMcO"
            gridtemplatecolumns=" minmax(100px, 0.4fr) minmax(100px, 0.5fr) minmax(100px, 0.5fr) minmax(100px, 0.4fr)"
            style="grid-template-columns: minmax(100px, 0.4fr) minmax(100px, 0.5fr) minmax(100px, 0.5fr) minmax(100px, 0.4fr);"
          >
            <th
              class="sc-hjmm4m-0 iOyNiA"
              tablesize="M"
            >
              <button
                class="sc-hjmm4m-2 kpjakb"
                tablesize="M"
              >
                <span
                  class="sc-1ntgl3l-0 ghpJwn"
                  textalign="initial"
                  textcolor="inherit"
                >
                  Date
                </span>
                <svg
                  class="sc-1pf3x8a-0 bMQNVa"
                  fill="none"
                  height="1em"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11.229 6.935a1 1 0 0 1 1.543 0l2.208 2.679a1 1 0 0 1-.77 1.636H9.79a1 1 0 0 1-.77-1.636l2.209-2.679zM12.771 17.065a1 1 0 0 1-1.543 0L9.02 14.386a1 1 0 0 1 .772-1.636h4.418a1 1 0 0 1 .771 1.636l-2.209 2.679z"
                    fill="#435465"
                  />
                </svg>
              </button>
              <span
                class="sc-hjmm4m-1 cuEDpa"
              />
            </th>
            <th
              class="sc-hjmm4m-0 iOyNiA"
              tablesize="M"
            >
              <button
                class="sc-hjmm4m-2 kpjakb"
                tablesize="M"
              >
                <span
                  class="sc-1ntgl3l-0 ghpJwn"
                  textalign="initial"
                  textcolor="inherit"
                >
                  Feedback From
                </span>
              </button>
              <span
                class="sc-hjmm4m-1 cuEDpa"
              />
            </th>
            <th
              class="sc-hjmm4m-0 iOyNiA"
              tablesize="M"
            >
              <button
                class="sc-hjmm4m-2 kpjakb"
                tablesize="M"
              >
                <span
                  class="sc-1ntgl3l-0 ghpJwn"
                  textalign="initial"
                  textcolor="inherit"
                >
                  Feedback Type
                </span>
              </button>
              <span
                class="sc-hjmm4m-1 cuEDpa"
              />
            </th>
            <th
              class="sc-hjmm4m-0 iOyNiA"
              tablesize="M"
            >
              <button
                class="sc-hjmm4m-2 kpjakb"
                tablesize="M"
              >
                <span
                  class="sc-1ntgl3l-0 ghpJwn"
                  textalign="initial"
                  textcolor="inherit"
                >
                  Action
                </span>
              </button>
              <span
                class="sc-hjmm4m-1 cuEDpa"
              />
            </th>
          </tr>
        </thead>
        <tbody
          class="sc-c7cniy-0 kShdfA"
        >
          <tr
            class="sc-1yxujaa-0 bgyIwA"
            gridtemplatecolumns=" minmax(100px, 0.4fr) minmax(100px, 0.5fr) minmax(100px, 0.5fr) minmax(100px, 0.4fr)"
            style="grid-template-columns: minmax(100px, 0.4fr) minmax(100px, 0.5fr) minmax(100px, 0.5fr) minmax(100px, 0.4fr);"
            tabindex="0"
          >
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
          </tr>
          <tr
            class="sc-1yxujaa-0 bgyIwA"
            gridtemplatecolumns=" minmax(100px, 0.4fr) minmax(100px, 0.5fr) minmax(100px, 0.5fr) minmax(100px, 0.4fr)"
            style="grid-template-columns: minmax(100px, 0.4fr) minmax(100px, 0.5fr) minmax(100px, 0.5fr) minmax(100px, 0.4fr);"
            tabindex="1"
          >
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
          </tr>
          <tr
            class="sc-1yxujaa-0 bgyIwA"
            gridtemplatecolumns=" minmax(100px, 0.4fr) minmax(100px, 0.5fr) minmax(100px, 0.5fr) minmax(100px, 0.4fr)"
            style="grid-template-columns: minmax(100px, 0.4fr) minmax(100px, 0.5fr) minmax(100px, 0.5fr) minmax(100px, 0.4fr);"
            tabindex="2"
          >
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
          </tr>
          <tr
            class="sc-1yxujaa-0 bgyIwA"
            gridtemplatecolumns=" minmax(100px, 0.4fr) minmax(100px, 0.5fr) minmax(100px, 0.5fr) minmax(100px, 0.4fr)"
            style="grid-template-columns: minmax(100px, 0.4fr) minmax(100px, 0.5fr) minmax(100px, 0.5fr) minmax(100px, 0.4fr);"
            tabindex="3"
          >
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
          </tr>
          <tr
            class="sc-1yxujaa-0 bgyIwA"
            gridtemplatecolumns=" minmax(100px, 0.4fr) minmax(100px, 0.5fr) minmax(100px, 0.5fr) minmax(100px, 0.4fr)"
            style="grid-template-columns: minmax(100px, 0.4fr) minmax(100px, 0.5fr) minmax(100px, 0.5fr) minmax(100px, 0.4fr);"
            tabindex="4"
          >
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
            <td
              class="sc-1e51eaa-0 hFOqbY"
              tablesize="M"
            >
              <div
                class="sc-15vbyef-0 bHPECM"
              />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </main>
</div>
`;
